qiskit/transpiler/passes/routing/algorithms/bip_model.py

```diff
@@ -0,0 +1,403 @@
+            TranspilerError: If size of virtual qubits and physical qubits differ, or
+                if coupling_map is not symmetric (bidirectional).
+        """
+        if not HAS_DOCPLEX:
+            raise MissingOptionalLibraryError(
+                libname="DOcplex",
+                name="Decision Optimization CPLEX Modeling for Python",
+                pip_install="pip install docplex",
+            )
+
+        if len(dag.qubits) != coupling_map.size():
+            raise TranspilerError(
+                "BIPMappingModel assumes the same size of virtual and physical qubits."
+            )
+        if not coupling_map.is_symmetric:
+            raise TranspilerError(
+                "BIPMappingModel assumes the coupling_map is symmetric (bidirectional)."
+            )
+
+        self._dag = dag
+        self._coupling = coupling_map
+
+        self.problem = None
+        self.solution = None
+        self._num_vqubits = len(self._dag.qubits)
+        self._num_pqubits = self._coupling.size()
+        self._arcs = self._coupling.get_edges()
+
+        # pylint: disable=unnecessary-comprehension
+        initial_layout = Layout.generate_trivial_layout(*dag.qregs.values())
+        self._virtual_to_index = {v: i for i, v in enumerate(initial_layout.get_virtual_bits())}
+        self._index_to_virtual = {i: v for i, v in enumerate(initial_layout.get_virtual_bits())}
+
+        # Construct internal circuit model
+        # Extract layers with 2-qubit gates
+        self._to_su4layer = []
+        self.su4layers = []
+        for lay in dag.layers():
+            laygates = []
+            subdag = lay["graph"]
+            for node in subdag.two_qubit_ops():
+                i1 = self._virtual_to_index[node.qargs[0]]
+                i2 = self._virtual_to_index[node.qargs[1]]
+                laygates.append((i1, i2))
+            if laygates:
+                self._to_su4layer.append(len(self.su4layers))
+                self.su4layers.append(laygates)
+            else:
+                self._to_su4layer.append(-1)
+        # Add dummy time steps inbetween su4layers. Dummy time steps can only contain SWAPs.
+        self.gates = []  # layered 2q-gates with dummy steps
+        for k, lay in enumerate(self.su4layers):
+            self.gates.append(lay)
+            if k == len(self.su4layers) - 1:  # do not add dummy steps after the last layer
+                break
+            self.gates.extend([[]] * dummy_timesteps)
+
+        logger.info("Num virtual qubits: %d", self.num_vqubits)
+        logger.info("Num physical qubits: %d", self.num_pqubits)
+        logger.info("Model depth: %d", self.depth)
+        logger.info("Dummy steps: %d", dummy_timesteps)
+
+    @property
+    def num_vqubits(self):
+        """Number of virtual qubits."""
+        return self._num_vqubits
+
+    @property
+    def num_pqubits(self):
+        """Number of physical qubits."""
+        return self._num_pqubits
```

Comment:
What's the advantage of having `@property`s here?
```suggestion
        self.num_vqubits = len(self._dag.qubits)
        self.num_pqubits = self._coupling.size()
        self._arcs = self._coupling.get_edges()

        # pylint: disable=unnecessary-comprehension
        initial_layout = Layout.generate_trivial_layout(*dag.qregs.values())
        self._virtual_to_index = {v: i for i, v in enumerate(initial_layout.get_virtual_bits())}
        self._index_to_virtual = {i: v for i, v in enumerate(initial_layout.get_virtual_bits())}

        # Construct internal circuit model
        # Extract layers with 2-qubit gates
        self._to_su4layer = []
        self.su4layers = []
        for lay in dag.layers():
            laygates = []
            subdag = lay["graph"]
            for node in subdag.two_qubit_ops():
                i1 = self._virtual_to_index[node.qargs[0]]
                i2 = self._virtual_to_index[node.qargs[1]]
                laygates.append((i1, i2))
            if laygates:
                self._to_su4layer.append(len(self.su4layers))
                self.su4layers.append(laygates)
            else:
                self._to_su4layer.append(-1)
        # Add dummy time steps inbetween su4layers. Dummy time steps can only contain SWAPs.
        self.gates = []  # layered 2q-gates with dummy steps
        for k, lay in enumerate(self.su4layers):
            self.gates.append(lay)
            if k == len(self.su4layers) - 1:  # do not add dummy steps after the last layer
                break
            self.gates.extend([[]] * dummy_timesteps)

        logger.info("Num virtual qubits: %d", self.num_vqubits)
        logger.info("Num physical qubits: %d", self.num_pqubits)
        logger.info("Model depth: %d", self.depth)
        logger.info("Dummy steps: %d", dummy_timesteps)

```

Reply:
Ah, fair enough. I was not consistent here. I meant to use @property to prevent accidental "set"  and tell users these are public properties clearly (in docstring) here, however, I didn't go for other public properties such as `problem`, `su4layers` etc. I'll accept your suggested simpler code because this is Python and this class is not a primary user-facing class, so simpler would be better.